alertingest/bin/alertdb_ingester.py

```diff
@@ -0,0 +1,71 @@
+import asyncio
+import argparse
+import os
+
+from alertingest.ingester import KafkaConnectionParams, IngestWorker
+from alertingest.storage import GoogleObjectStorageBackend
+from alertingest.schema_registry import SchemaRegistryClient
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        "alertdb-ingester",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+        description="Run a worker to copy alerts from Kafka into an object store backend.",
+    )
+    parser.add_argument(
+        "--gcp-project",
+        type=str,
+        default="alert-stream",
+        help="when using the google-cloud backend, the name of the GCP project",
+    )
+    parser.add_argument(
+        "--gcp-bucket",
+        type=str,
+        default="rubin-alert-archive",
+        help="when using the google-cloud backend, the name of the Google Cloud Storage bucket",
+    )
+    parser.add_argument(
+        "--kafka-host",
+        type=str,
+        default="alertbroker-scratch.lsst.codes",
+        help="kafka host with alert data",
+    )
+    parser.add_argument(
+        "--kafka-topic",
+        type=str,
+        default="alerts",
+        help="name of the Kafka topic with alert data",
+    )
+    parser.add_argument(
+        "--kafka-group",
+        type=str,
+        default="alertdb-ingester",
+        help="Name of a Kafka Consumer group to run under",
+    )
+    parser.add_argument(
+        "--kafka-username",
+        type=str,
+        default="admin",
+        help="Username to use when connecting to Kafka",
+    )
+    parser.add_argument(
+        "--schema-registry-host",
+        type=str,
+        default="alertschemas-scratch.lsst.codes",
+        help="Address of a Confluent Schema Registry server hosting schemas",
+    )
+    args = parser.parse_args()
+
+    kafka_params = KafkaConnectionParams(
+        args.kafka_host,
+        args.kafka_topic,
+        args.kafka_group,
+        args.kafka_username,
+        os.environ["ALERTDB_KAFKA_PASSWORD"],
+    )
+    backend = GoogleObjectStorageBackend(args.gcp_project, args.gcp_bucket)
+    registry = SchemaRegistryClient(args.schema_registry_host)
+
+    worker = IngestWorker(kafka_params, backend, registry)
+    asyncio.get_event_loop().run_until_complete(worker.run())
```

Comment:
Is there anything about the app that's asynchronous? Or is the asyncio use mostly to be able to use async libraries like aiokafka and kafkit? I bring this up because if you end up finding the asyncio stuff harder to test or debug, then there might be no loss to switching to synchronous code and [kafka-python](https://kafka-python.readthedocs.io/en/master/index.html).

With aiokafka I occasionally get weird log messages about auto offset commits failing (which I don't think is relevant here, of course). Here's a page that weights aiokafka and kafka-python: https://aiokafka.readthedocs.io/en/stable/kafka-python_difference.html

Reply:
Ah, interesting. Yeah, nothing asynchronous here - we handle exactly one alert at a time.

I would like to handle multiple at once, though. The only coordination point between coroutines is probably committing offsets, so there's a lot of benefit to concurrently uploading several alerts at once. There might be races possible when doing the schema-existence checks, but races are OK, since uploading schemas (or populating the local schema cache) should be idempotent.

I think this might have a big impact on performance in production, since alerts are relatively large (~20-80KB) so we will spend a noticeable amount of time just shipping bytes out on the network - which could be spent doing other stuff.

The offset committing logic will need to be careful. It needs to only commit offset N when *all* of the messages (N-1, N-2, N-3... N-100) have successfully uploaded. This would seem to require a shared datastructure across the coroutines to keep track and know if any have raised exceptions.